synapse/lib/types.py

```diff
@@ -304,39 +311,74 @@ def __init__(self, tlib, name, **info):
         CompType.__init__(self, tlib, name, **info)
         self.sepr = info.get('sep',',')
         self.lower = info.get('lower',0)
+        self.recursive = info.get('recursive',0)
+        self.reverse = info.get('reverse',0)
 
     def norm(self, valu, oldval=None):
         reprs = []
 
         if self.lower:
             valu = valu.lower()
 
-        parts = valu.split(self.sepr,len(self.fields))
-        for part,(name,tobj) in self._zipvals(parts):
-            reprs.append( tobj.repr(tobj.parse(part)) )
-        return self.sepr.join(reprs)
+        if self.reverse:
+            parts = valu.rsplit(self.sepr,len(self.fields)-1)
+        else:
+            parts = valu.split(self.sepr,len(self.fields)-1)
+
+        if self.recursive:
+            for part,(name,tobj) in self._zipvals(parts):
+                if tobj == self:
+                    # do not recursively norm
+                    reprs.append(part)
+                else:
+                    reprs.append( tobj.repr(tobj.parse(part)) )
+            return self.sepr.join(reprs)
+
+        else:
+            for part,(name,tobj) in self._zipvals(parts):
+                reprs.append( tobj.repr(tobj.parse(part)) )
+            return self.sepr.join(reprs)
 
     def repr(self, valu):
         return valu
 
     def chop(self, valu):
         subs = {}
-        parts = valu.split(self.sepr,len(self.fields))
+        if self.reverse:
+            parts = valu.rsplit(self.sepr,len(self.fields)-1)
+        else:
+            parts = valu.split(self.sepr,len(self.fields)-1)
 
         if self.lower:
             valu = valu.lower()
 
         reprs = []
-        for part,(name,tobj) in self._zipvals(parts):
 
-            norm = tobj.parse(part)
-            norm,nsub = tobj.chop(norm)
+        if self.recursive:
```

Comment:
I feel like these cases are the same...  If it just always checks for tobj == self, we should be free and clear rather than having those same parse()/chop()/repr() blocks in 2 locations